pkg/appstate/wait.go

```diff
@@ -0,0 +1,228 @@
+			state := CalculateIngressState(clientset, r)
+			if state == types.StateReady {
+				return nil
+			}
+		}
+
+		time.Sleep(WaitForResourceInterval)
+	}
+}
+
+func WaitForPersistentVolumeClaimToBeReady(clientset kubernetes.Interface, namespace, name string) error {
+	for {
+		r, err := clientset.CoreV1().PersistentVolumeClaims(namespace).Get(context.TODO(), name, metav1.GetOptions{})
+		if err != nil && !kuberneteserrors.IsNotFound(err) {
+			return errors.Wrap(err, "failed to get existing persistentvolumeclaim")
+		}
+
+		if !kuberneteserrors.IsNotFound(err) {
+			state := CalculatePersistentVolumeClaimState(r)
+			if state == types.StateReady {
+				return nil
+			}
+		}
+
+		time.Sleep(WaitForResourceInterval)
+	}
+}
+
+func WaitForServiceToBeReady(clientset kubernetes.Interface, namespace, name string) error {
+	for {
+		r, err := clientset.CoreV1().Services(namespace).Get(context.TODO(), name, metav1.GetOptions{})
+		if err != nil && !kuberneteserrors.IsNotFound(err) {
+			return errors.Wrap(err, "failed to get existing service")
+		}
+
+		if !kuberneteserrors.IsNotFound(err) {
+			state := CalculateServiceState(clientset, r)
+			if state == types.StateReady {
+				return nil
+			}
+		}
+
+		time.Sleep(WaitForResourceInterval)
+	}
+}
+
+func WaitForStatefulSetToBeReady(clientset kubernetes.Interface, namespace, name string) error {
+	for {
+		r, err := clientset.AppsV1().StatefulSets(namespace).Get(context.TODO(), name, metav1.GetOptions{})
+		if err != nil && !kuberneteserrors.IsNotFound(err) {
+			return errors.Wrap(err, "failed to get existing statefulset")
+		}
+
+		if !kuberneteserrors.IsNotFound(err) {
+			state := CalculateStatefulSetState(clientset, namespace, r)
+			if state == types.StateReady {
+				return nil
+			}
+		}
+
+		time.Sleep(WaitForResourceInterval)
+	}
+}
+
+func WaitForGenericResourceToBeReady(dr dynamic.ResourceInterface, name string) (err error) {
+	for {
+		_, err := dr.Get(context.TODO(), name, metav1.GetOptions{})
+		if err != nil && !kuberneteserrors.IsNotFound(err) {
+			return errors.Wrap(err, "failed to get existing resource")
+		}
+
+		if !kuberneteserrors.IsNotFound(err) {
+			return nil
+		}
+
+		time.Sleep(WaitForResourceInterval)
+	}
+}
+
+func WaitForProperty(namespace, name string, gvk *schema.GroupVersionKind, path, desiredValue string) error {
+	dr, err := k8sutil.GetDynamicResourceInterface(gvk, namespace)
+	if err != nil {
+		return errors.Wrap(err, "failed to get dynamic resource interface")
+	}
+
+	for {
+		r, err := dr.Get(context.TODO(), name, metav1.GetOptions{})
+		if err != nil && !kuberneteserrors.IsNotFound(err) {
+			return errors.Wrap(err, "failed to get existing resource")
+		}
+
+		if !kuberneteserrors.IsNotFound(err) {
+			if matches, err := resourcePropertyMatchesValue(r, path, desiredValue); err != nil {
+				return errors.Wrap(err, "failed to check if resource property matches value")
+			} else if matches {
+				return nil
+			}
+		}
+
+		time.Sleep(WaitForResourceInterval)
```

Comment:
is there a timeout? can it be configured?

Reply:
no timeout as it stands now, but is something we have been discussing

Reply:
we're sticking with no timeout for now as we don't currently have any concept of timeouts during the deployment of native k8s manifests, so we will likely want to look at it more holistically outside of just this feature.  does sound like something that would not be difficult to implement here, so I could see that happening if there is enough demand for it